i18n: Add reactive moment and formatted date components

blowery opened the pull request:
Add two new components to support loading moment locale's async. First, a `withMoment()` HoC that watches for changes to the current locale and passes a localized moment to the wrapped component. Second, a `FormattedDate` component that uses `withMoment` to provide localized, formatted dates.

To test, visit http://calypso.localhost:3000/devdocs/design/formatted-date and play around. Try changing the locale and the date and the format string.

blowery committed:
i18n: Add reactive moment and formatted date components

Add two new components to support loading moment locale's async. First, a withMoment HoC that watches for changes to the current locale and passes a localized moment to the wrapped component. Second, a FormattedDate component that uses withMoment to provide localized, formatted dates.

blowery commented:
This is a different take on https://github.com/Automattic/wp-calypso/pull/26173 which would allow us to reactively update things that use moment when the locale changes and we need to switch translations.

blowery commented:
Part of me really wants to be able to say `<FormattedDate date={ someDate } format="ll" locale="es" />` but I have no real use case for it.

flootr commented on client/components/with-localized-moment/index.js:
```diff
@@ -0,0 +1,81 @@
+				return;
+			}
+			if ( this.props.locale !== 'en' ) {
+				const loadingLocale = this.props.locale;
```
Do we need some kind of whitelist or fallback for this? This only came to mind after playing around in DevDocs and using a non-existent locale and may not be applicable to the real use case (where afaik the user chooses from a list of available locales).

blowery replied:
Good idea!

flootr replied:
Or maybe a `.catch()` which loads `en` in this case? 🤔

jsnajdr commented:
I posted a few nit comments, but generally this PR looks good and should work reasonably well.

There is a can of worms called SSR. Can server synchronously render a `FormattedDate` in the correct locale? Can client rehydrate the UI on initial render with exactly the same content? This is a tough problem, see https://github.com/Automattic/wp-calypso/pull/26495#issuecomment-421943817 for more context.

I also don't like how the component instances race with each other to load and update the locale. But the middleware in #26173 doesn't seem to have a correct solution: the instances don't get rerendered after the locale chunk finishes loading and after `moment.locale( newLocale )` is called.

React context could be a good solution:
- context provider that lives right under the Redux provider, listens for Redux locale change and is responsible for the locale load.
- `FormattedDate` instances being a context consumer reacting to context change.

blowery commented:
@jsnajdr using a context sounds sane... I think you're more familiar with how those work than I am, would you have time to try and code that up?

jsnajdr commented:
> linkTag.setAttribute("data-webpack", true);

I can try 🙂

jsnajdr reviewed (commented):
Oops, I see I forgot to submit a review yesterday 😕

sirreal commented on client/components/with-localized-moment/index.js:
```diff
@@ -0,0 +1,81 @@
+
+function getDisplayName( WrappedComponent ) {
+	return WrappedComponent.displayName || WrappedComponent.name || 'Component';
+}
```
How about `createHigherOrderComponent` from `@wordpress/compose` 😁 

https://github.com/WordPress/gutenberg/blob/3b6c167cab50e77c44b21c74e3dd2bdf8b890ff4/packages/compose/src/create-higher-order-component/index.js

jsnajdr replied:
Of course, it's right here under my nose 😄

blowery committed:
use withMoment HoC to supply moment instance

blowery committed:
use localized moment hoc

blowery committed:
use moment from withMoment hoc

blowery committed:
switch to withMoment hoc

blowery committed:
extract a small tool to convert a moment from any locale to the current locale

blowery commented:
Things that use moment.format in non-reactive ways, mostly using moment instances that are passed in.
state/stats/lists/utils.js
my-sites/domains/domain-management/edit/site-redirect.jsx
my-sites/domains/domain-management/edit/mapped-domain.jsx
my-sites/domains/domain-management/edit/registered-domain.jsx
my-sites/domains/domain-management/edit/transfer.jsx
my-sites/domains/components/domain-warnings/index.jsx
my-sites/plans/current-plan/header.jsx
lib/human-date/index.js

jsnajdr committed:
Pass correct value (moment obj rather than locale str) to ctx provider

jsnajdr commented on client/components/with-localized-moment/context.js:
```diff
@@ -0,0 +1,54 @@
+	}
+
+	render() {
+		return <Provider value={ moment }>{ this.props.children }</Provider>;
```
This code was passing the `currentLocale` string as the context value rather than the `moment` object. Fixed in https://github.com/Automattic/wp-calypso/pull/26476/commits/30b3b967ef09ead9c51da5f00025d4bfcdf249bb

blowery replied:
@jsnajdr I think that was on purpose. The moment instance doesn't change, so it doesn't always cause the re-render necessary when the locale changes.

jsnajdr replied:
But the component wrapped by `withMoment()` needs to receive a moment instance, not a locale string. Calypso crashed whenever the `moment` prop from `withMoment()` was used. For example, in the "Joined ..." label in `/me/account`

<img width="433" alt="screen shot 2018-10-31 at 15 21 24" src="https://user-images.githubusercontent.com/664258/47794434-bc247d00-dd20-11e8-9ed2-04c48f8c7c50.png">

I'm trying another trick in https://github.com/Automattic/wp-calypso/pull/26476/commits/e16769881266f13e60ccf75444e16bb549e3f62a: set a `{ moment }` object as the value of the context instead of just the constant `moment`. The container object is in local state and updated on locale change. Inspired by https://reactjs.org/docs/context.html#caveats

The only way to break this is to put a pure component between `withMoment` and the site where the `moment` prop is used:
```jsx
class Label extends PureComponent {
  render() {
    return <span>{ this.props.moment( this.props.date ).format() }</span>
  }
}

class LocalizedLabel = withMoment( Label );

<Label date="2018-10-31" />
```

The `LocalizedComponent` itself will rerender on locale change, but the inner `Label` will receive unchanged `moment` and `date` props! Because it's pure, it won't rerender.

Several possible solutions if this turns out to really be a problem:
1. Create a real new instance of `moment` on locale change. Is there any caveat that prevents that?
2. Pass an additional helper `momentLocale` string prop to the child of `withMoment`, with the purpose to get  through `PureComponent` defenses.
3. Instead of relying on context updates, the Moment Provider would be an event emitter and Moment Consumers can subscribe to updates.

blowery replied:
I don't think 1. is possible, as moment itself is a singleton

I like 2 the best.

blowery replied:
updated to 2 in https://github.com/Automattic/wp-calypso/pull/26476/commits/04a49d56dda0a358ae792c0f9f6ada09b017ad15

jsnajdr commented:
> Things that use moment.format in non-reactive ways, mostly using moment instances that are passed in.

@blowery Do you think any of the listed issues block this PR? My impression is that they already didn't react to locale changes properly and will continue to not-react the same way even after the PR. Can we create an issue with a task list to fix them one by one?

Some fixes with require removing Moment objects from Redux state in favor of primitive representations (Unix time integers, ISO strings).

Some are just date comparisons and manipulations that should be locale independent. We only need to verify and test carefully.

Some files are rabbit holes because of the ESLint warnings. That happens in Domains. Some Domains components still rely on mixins and need to be converted into ES6 classes first.

jsnajdr committed:
Set Redux UI locale state in Quick Language Picker

After changing language in Quick Language Picker, set `ui.language.localeSlug` in Redux
in addition to updating the `i18n-calypso` internal state.

Lets components that use that state bit (like the new `withMoment`) react to locale changes.

jsnajdr committed:
Moment Context: set new value of context on each locale change

The `moment` object itself never changes and a context update is not triggered on
locale change. This patch sets a new context value on locale change, triggering
rerenders as needed.

See also https://reactjs.org/docs/context.html#caveats

blowery committed:
Pass the current lang slug to the child to break purity

blowery commented on client/components/with-localized-moment/index.js:
```diff
@@ -0,0 +1,81 @@
+		.finally( () => {
+			momentLoaders.delete( locale );
+		} );
+	momentLoaders.set( locale, newLoader );
```
Hrm. This is happening again in the current code. We end up calling moment.locale once for every mounted moment.provider.

blowery replied:
fixed in https://github.com/Automattic/wp-calypso/pull/26476/commits/25ea41aed0f0a5e57f54e6fdea60f84d0ac18f75

not sure it really has much impact though. saves one render.

blowery committed:
change prop name to momentLocale to prevent collisions, only change locale and set state if it changed

blowery committed:
move to a functional component for FormattedDate

blowery commented on client/components/formatted-date/index.js:
```diff
@@ -0,0 +1,23 @@
+ */
+import withMoment from 'components/with-localized-moment';
+
+class FormattedDate extends Component {
```
done

jsnajdr reviewed (commented):
When writing Enzyme tests, I discovered two subtle bugs in the Moment Provider. I'm going to push a series of commits that add the test and also fix these bugs.

Also, there are a few places where we render a new React root with Redux Provider, but the Moment Provider is not there:
- the `RootChild` component
- the `renderWithReduxStore` helper

I'll add them there -- or am I missing something?

blowery commented on client/components/with-localized-moment/context.js:
```diff
@@ -0,0 +1,70 @@
+		// is moment set to the current locale?
+		if ( currentLocale === moment.locale() ) {
+			return;
+		}
```
Nice catch!

blowery replied:
I think we still need to fix this?

blowery replied:
oh it was, in https://github.com/Automattic/wp-calypso/pull/26476/commits/52818792c135cbf9ddc692b9b8b27138003aba63

jsnajdr commented on client/components/with-localized-moment/context.js:
```diff
@@ -0,0 +1,70 @@
+		// validate that someone else hasn't already changed the moment locale
+		if ( moment.locale() !== currentLocale ) {
+			moment.locale( currentLocale );
+			this.setState( {
```
If we do a repeated switch though, `en` -> `cs` -> `en` -> `cs`, then we need to explicitly call `moment.locale( 'cs' )` on the second `en` -> `cs` switch -- the locale is already loaded and the `moment.defineLocale` call inside the module won't happen again.

jsnajdr committed:
Expose a `loadingLocalePromise` property on the provider instance for tests

Tests need to wait until the load has finished and only then check for prop
updates and rerenders.

jsnajdr committed:
Add babel-plugin-dynamic-import-node to Babel config for tests: makes import() work

The plugin converts dynamic `import()` statements to promisified `require` calls that
Node.js understands. Allows us to test code that uses dynamic imports, namely the
async loading of moment.js locales.

jsnajdr committed:
Add Enzyme unit tests for the MomentProvider

jsnajdr committed:
Ensure that provider state is always updated correctly

Fixes the tests that are failing if this patch is not applied.

jsnajdr committed:
Added MomentProvider to RootChild and renderWithReduxStore

Two places where things get rendered into new React roots.

jsnajdr committed:
Webpack config: ignore moment locales only when building Calypso client

Let's not affect server and other users of Calypso SDK.

jsnajdr committed:
Fix image selector test by mocking the getCurrentLocaleSlug selector

jsnajdr commented:
To make the Desktop tests green, Automattic/wp-desktop#531 needs to be merged first. It adds a plugin to parse dynamic `import()` statements to its Babel config.

jsnajdr reviewed (approved):
Everything is 🍏 now and I believe we're ready to merge. What do you think @blowery? Or shall we seek an independent third party to review some more?

blowery commented:
@jsnajdr I say land it and see if any issues crop up.

jsnajdr commented:
OK I'll merge in the morning and will watch if there is any trouble.
